### PR TITLE
Build compile_commands.json from xcode project (#654)

### PR DESCRIPTION
**Pull request by guinzoo**
I don't know what is the best way to check input file argument. Feel free to refactor it.

### CODE REVIEW

**Comment by niosus** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/zabugr/EasyClangComplete/pullRequest?prid=4707417&bid=15724981)

File: EasyClangComplete.py [hunk 1]
```diff
@@ -183,6 +183,44 @@ def run(self, edit):
...
 
+class EccGenerateXcodeCommandsCommand(sublime_plugin.TextCommand):
+    """Command that generates compile_commands.json using xcodebuild and xcpretty"""
+
+    def run(self, edit):
+        settings = EasyClangComplete.settings_manager.settings_for_view(self.view)
+        project_file = self.view.window().project_file_name()
+        if settings.xcode_project and project_file:
+            project_dir = path.dirname(project_file)
+            os.chdir(project_dir)
```
Comment (niosus) [inline EasyClangComplete.py]: Codacy Issue found: [undefined name 'os'](https://app.codacy.com/app/zabugr/EasyClangComplete/pullRequest?prid=4707417)

File: plugin/utils/tools.py [hunk 2]
```diff
@@ -68,6 +68,17 @@ def run_command(command, shell=False, cwd=path.curdir, env=environ,
                 "Executable file not found executing: {}".format(command))
         return output_text
 
+    @staticmethod
+    def run_command_with_logging(command, cwd=path.curdir):
+        proc = subprocess.Popen([command], shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, cwd=cwd)
+        while proc.poll() is None:
+            try:
+                output = proc.stdout.readline().decode('UTF-8')
+                print(output, end='')
```
Comment (niosus) [inline plugin/utils/tools.py]: Codacy Issue found: [invalid syntax](https://app.codacy.com/app/zabugr/EasyClangComplete/pullRequest?prid=4707417)

File: plugin/flags_sources/compilation_db.py [hunk 3]
```diff
@@ -200,9 +200,13 @@ def filter_bad_arguments(argument_list):
                     skip_next = True
 
                 continue
-            if i == len(argument_list) - 1:
-                # ignore the last element as it is a file to compile, something
-                # like 'test.cpp'
+            if argument.endswith('.c'):
```
Comment (niosus) [inline plugin/flags_sources/compilation_db.py]: This is not enough as the extensions can be arbitrary. For example, Google tends to use `.cc`, etc.

File: plugin/utils/flag.py [hunk 4]
```diff
@@ -249,14 +249,14 @@ def build(self):
         "-dependency-dot",
         "-dependency-file",
         "-F",
-        "-fmodules-cache-path",
```
Comment (niosus) [inline plugin/utils/flag.py]: Why is this removed?

File: plugin/utils/flag.py [hunk 5]
```diff
@@ -275,6 +275,7 @@ def build(self):
         "-mthread-model",
         "-o",
         "-serialize-diagnostics",
+        "--serialize-diagnostics",
```
Comment (niosus) [inline plugin/utils/flag.py]: We either would have to add all with `--` or none. I am not a fan of having one flag different from other.

File: plugin/utils/tools.py [hunk 6]
```diff
@@ -68,6 +69,17 @@ def run_command(command, shell=False, cwd=path.curdir, env=environ,
                 "Executable file not found executing: {}".format(command))
         return output_text
 
+    @staticmethod
+    def run_command_with_logging(command, cwd=path.curdir):
```
Comment (niosus) [inline plugin/utils/tools.py]: Why do we need this function?

File: plugin/utils/flag.py (hunk 4 above)
Comment (guinzoo) [reply plugin/utils/flag.py]: It should be used with equality (-fmodules-cache-path=<directory>)

File: plugin/utils/flag.py (hunk 5 above)
Comment (guinzoo) [reply plugin/utils/flag.py]: There is not much flags that support both prefixes. I have no idea how to do it better way.

File: plugin/utils/tools.py (hunk 6 above)
Comment (guinzoo) [reply plugin/utils/tools.py]: xcpretty parses output from xcodebuild, and to get commands for every file you should rebuild whole project. That's a long process, and to see what's going on and to be sure that all is fine it's better to print logs in real time.

File: plugin/flags_sources/compilation_db.py (hunk 3 above)
Comment (guinzoo) [reply plugin/flags_sources/compilation_db.py]: We can list all possible extensions or compare argument with file name/path.

**Comment by guinzoo** [comment]
I think you are right. I started it because had problems with your plugin. Now I have idea how to put all needed arguments for a project by hand.

File: plugin/flags_sources/compilation_db.py (hunk 3 above)
Comment (niosus) [reply plugin/flags_sources/compilation_db.py]: But why is the last element not the file to compile? What is the last element in such a line?

File: plugin/utils/flag.py (hunk 4 above)
Comment (niosus) [reply plugin/utils/flag.py]: It is already in the list of separable flags. It will be used with `=` sign. Or does this not work for you?

File: plugin/flags_sources/compilation_db.py (hunk 3 above)
Comment (guinzoo) [reply plugin/flags_sources/compilation_db.py]: Last element is output file, right after file to compile.

File: plugin/utils/flag.py (hunk 4 above)
Comment (guinzoo) [reply plugin/utils/flag.py]: It hadn't worked in my case, so I removed that line.

File: plugin/utils/tools.py (hunk 6 above)
Comment (niosus) [reply plugin/utils/tools.py]: I don't think that the console is the place to print things. If this is interfacing the user, we should think about showing a panel.

**Comment by niosus** [comment]
I'm sorry, but I probably will close this. At this point, there is a number of things I am not really happy with in the current implementation and I don't know enough to judge if they are actually valid. Also, I won't be able to support this behavior after it gets merged as I simply don't know enough about Objective C. Hope there is no hard feelings. I appreciate the effort a lot, but I don't see a sustainable way to merge this right now.
